File: thinking-categorically/cartesian-closed-category/pure-subcategory/pure.ts
// Terminal Object: void/undefined (unit type)
type Terminal = void;
const terminal: Terminal = undefined;

// Currying: ((A × B) => C) → (A => (B => C))
const curry = <A, B, C>(f: (pair: Product<A, B>) => C): (x: A) => (y: B) => C =>
  (x: A) => (y: B) => f([x, y]);

// Uncurrying: (A => (B => C)) → ((A × B) => C)
const uncurry = <A, B, C>(f: (x: A) => (y: B) => C): (pair: Product<A, B>) => C =>
  ([x, y]: Product<A, B>) => f(x)(y);

// Every type has exactly one function to terminal
const toTerminal = <A>(x: A): Terminal => undefined;

// Binary Products: Tuple types [A, B]
type Product<A, B> = [A, B];

// Product construction and projections
const makePair = <A, B>(x: A, y: B): Product<A, B> => [x, y];

const fst = <A, B>([x, _]: Product<A, B>): A => x;
const snd = <A, B>([_, y]: Product<A, B>): B => y;

// Universal property: pairing morphism
const pairing = <X, A, B>(
  f: (x: X) => A,
  g: (x: X) => B
): (x: X) => Product<A, B> =>
  (x: X) => [f(x), g(x)];

// Exponential Objects: Function types (A) => B
type Exponential<A, B> = (x: A) => B;

// Evaluation morphism
const evalMorphism = <A, B>(func: Exponential<A, B>, arg: A): B => func(arg);

// Alternative eval that takes a product
const evalProduct = <A, B>([func, arg]: Product<Exponential<A, B>, A>): B =>
  func(arg);

// Example 1: Terminal object and product projections
function example1() {
  const unitResult: Terminal = terminal;               // unique inhabitant
  const dropped: Terminal = toTerminal("anything");   // morphism to 1
  const pairAB = makePair<number, string>(42, "answer");
  const firstOfPair = fst(pairAB);                     // π₁
  const secondOfPair = snd(pairAB);                    // π₂
  console.log("Example 1 (terminal, products):", { unitResult, dropped, pairAB, firstOfPair, secondOfPair });
}

// Example 2: Universal property of products (pairing) + evaluation
function example2() {
  const build = pairing((x: number) => x + 1, (x: number) => x * 2); // ⟨f,g⟩
  const from5 = build(5); // [6,10]
  const incr: Exponential<number, number> = x => x + 1;
  const evaluated = evalMorphism(incr, 10);                 // eval(f, a)
  const evaluatedViaProduct = evalProduct([incr, 10]);      // eval ∘ pairing
  console.log("Example 2 (pairing, eval):", { from5, evaluated, evaluatedViaProduct });
}

// Example 3: Currying/Uncurrying isomorphism (exponential structure)
function example3() {
  const addPair = ([a,b]: Product<number, number>) => a + b;          // (A × B) → C
  const curried = curry(addPair);                                     // A → (B → C)
  const uncurried = uncurry(curried);                                 // round trip
  const roundTrip = uncurried([3,4]);                                 // 7
  // Partial application exhibits internal hom
  const add10 = curried(10);                                          // B → C
  const fifteen = add10(5);                                           // 15
  console.log("Example 3 (curry/uncurry):", { roundTrip, fifteen });
}

// Example 4: Higher-order functions and composition (exponential structure)
const compose = <A, B, C>(
  f: Exponential<B, C>,
  g: Exponential<A, B>
): Exponential<A, C> =>
  (x: A) => f(g(x));

function example4() {
  const processNumber: Exponential<number, string> = compose(
    compose(String, (x: number) => x * 2), // (*2) then to string
    (x: number) => x + 1                   // (+1) first
  );
  const result = processNumber(5); // "12"
  console.log("Example 4 (composition / pipeline):", { result });
}

// Example 5: Familiar higher-order operations as exponentials
function example5() {
  const numbers = [1, 2, 3, 4, 5];
  const doubled = numbers.map((x: number) => x * 2);        // map: functorial action
  const evens = numbers.filter((x: number) => x % 2 === 0); // predicate as exponential
  const sum = numbers.reduce((acc: number, x: number) => acc + x, 0); // fold uses evaluation repeatedly
  console.log("Example 5 (arrays & exponentials):", { doubled, evens, sum });
}

// Driver to run all examples in order
function runAllExamples() {
  example1();
  example2();
  example3();
  example4();
  example5();
}

runAllExamples();

File: thinking-categorically/cartesian-closed-category/pure-subcategory/pure.cs
using System;

// Terminal Object: Unit struct (equivalent to Haskell's ())
public struct Unit
{
    public static readonly Unit Value = new Unit();

    // Every type has exactly one function to Unit
    public static Unit ToTerminal<T>(T value) => Value;
}

// Binary Products: Generic tuple
public struct Product<A, B>
{
    public A First { get; }
    public B Second { get; }

    public Product(A first, B second)
    {
        First = first;
        Second = second;
    }

    // Projections
    public static A Fst(Product<A, B> product) => product.First;
    public static B Snd(Product<A, B> product) => product.Second;
}

// Exponential Objects: Generic delegates
public delegate TResult Exponential<T, TResult>(T input);

public static class CartesianClosedCategory
{
    // Product construction
    public static Product<A, B> MakePair<A, B>(A first, B second) =>
        new Product<A, B>(first, second);

    // Universal property: pairing morphism
    public static Func<X, Product<A, B>> Pairing<X, A, B>(
        Func<X, A> f,
        Func<X, B> g) =>
        x => new Product<A, B>(f(x), g(x));

    // Evaluation morphism
    public static TResult Eval<T, TResult>(
        Exponential<T, TResult> func,
        T argument) =>
        func(argument);

    // Alternative eval that takes a product
    public static TResult EvalProduct<T, TResult>(
        Product<Exponential<T, TResult>, T> pair) =>
        pair.First(pair.Second);

    // Currying: ((A × B) → C) → (A → (B → C))
    public static Func<A, Func<B, C>> Curry<A, B, C>(
        Func<Product<A, B>, C> f) =>
        a => b => f(new Product<A, B>(a, b));

    // Uncurrying: (A → (B → C)) → ((A × B) → C)
    public static Func<Product<A, B>, C> Uncurry<A, B, C>(
        Func<A, Func<B, C>> f) =>
        pair => f(pair.First)(pair.Second);

    // Function composition
    public static Exponential<A, C> Compose<A, B, C>(
        Exponential<B, C> f,
        Exponential<A, B> g) =>
        x => f(g(x));
}

class Program
{
    static void Main()
    {
        // Original function: Product<int, int> -> int
        Func<Product<int, int>, int> multiply =
            pair => pair.First * pair.Second;

        // Curry it: int -> (int -> int)
        var curriedMultiply = CartesianClosedCategory.Curry(multiply);

        // Uncurry it back: Product<int, int> -> int
        var uncurriedMultiply = CartesianClosedCategory.Uncurry(curriedMultiply);

        // Test the isomorphism
        var testPair = new Product<int, int>(6, 7);

        Console.WriteLine($"Original: {multiply(testPair)}");           // 42
        Console.WriteLine($"Curried: {curriedMultiply(6)(7)}");        // 42
        Console.WriteLine($"Uncurried: {uncurriedMultiply(testPair)}"); // 42

        // Partial application demonstrates exponential objects
        var multiplyBy5 = curriedMultiply(5);
        Console.WriteLine($"5 × 8 = {multiplyBy5(8)}"); // 40

        // Higher-order functions
        Exponential<int, int> addOne = x => x + 1;
        Exponential<int, int> double_ = x => x * 2;
        Exponential<int, string> toString = x => x.ToString();

        // Function composition
        var pipeline = CartesianClosedCategory.Compose(
            toString,
            CartesianClosedCategory.Compose(double_, addOne)
        );

        Console.WriteLine($"Pipeline(5) = {pipeline(5)}"); // "12"

        // Evaluation morphism
        Console.WriteLine($"Eval(addOne, 10) = {CartesianClosedCategory.Eval(addOne, 10)}"); // 11

        // Using pairing morphism
        var coordPairing = CartesianClosedCategory.Pairing<int, int, int>(
            x => x + 1,    // f: int -> int
            x => x * 2     // g: int -> int
        );

        var coords = coordPairing(5);
        Console.WriteLine($"Pairing(5) = ({coords.First}, {coords.Second})"); // (6, 10)
    }
}